lib/Dialect/Torch/IR/TorchOps.cpp

```diff
@@ -2043,7 +2043,13 @@ void PrimTupleUnpackOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
     if (!tupleConstruct)
       return failure();
 
-    rewriter.replaceOp(op, tupleConstruct.getElements());
+    llvm::SmallVector<Value> derefinedElements;
+    for (auto [type, element] :
+         llvm::zip(op.getResultTypes(), tupleConstruct.getElements())) {
+      derefinedElements.push_back(
+          rewriter.createOrFold<DerefineOp>(op.getLoc(), type, element));
```

Comment:
We need to maintain the correct type for the value.

IIRC, there is a TODO somewhere in the MLIR pattern rewriter to re-enable a check forcing rewrites to maintain exact types. That would have caught this.

Let me know if `DerefineOp` is not the right op for the type conversion. I took it from what I was seeing in our model.